esmvaltool/utils/recipe_filler.py

```diff
@@ -0,0 +1,924 @@
+
+Caveats:
+
+- the tool doesn't yet work for derived variables;
+- operation restricted to CMIP data.
+
+Have fun!
+"""
+import argparse
+import datetime
+import itertools
+import logging
+import logging.config
+import os
+import shutil
+import time
+import warnings
+from glob import glob
+from pathlib import Path
+
+import yaml
+from ruamel.yaml import YAML
+
+import esmvalcore
+from esmvalcore.cmor.table import CMOR_TABLES, read_cmor_tables
+
+logger = logging.getLogger(__name__)
+
+CFG = {}
+
+
+# standard libs from esmvalcore ported here to avoid private func import
+def load_config_developer(cfg_file=None):
+    """Load the config developer file and initialize CMOR tables."""
+    cfg_developer = read_config_developer_file(cfg_file)
+    for key, value in cfg_developer.items():
+        CFG[key] = value
+    read_cmor_tables(CFG)
+
+
+def _purge_file_handlers(cfg: dict) -> None:
+    """Remove handlers with filename set.
+
+    This is used to remove file handlers which require an output
+    directory to be set.
+    """
+    cfg['handlers'] = {
+        name: handler
+        for name, handler in cfg['handlers'].items()
+        if 'filename' not in handler
+    }
+    prev_root = cfg['root']['handlers']
+    cfg['root']['handlers'] = [
+        name for name in prev_root if name in cfg['handlers']
+    ]
+
+
+def _update_stream_level(cfg: dict, level=None):
+    """Update the log level for the stream handlers."""
+    handlers = cfg['handlers']
+
+    for handler in handlers.values():
+        if level is not None and 'stream' in handler:
+            if handler['stream'] in ('ext://sys.stdout', 'ext://sys.stderr'):
+                handler['level'] = level.upper()
+
+
+def _get_log_files(cfg: dict, output_dir: str = None) -> list:
+    """Initialize log files for the file handlers."""
+    log_files = []
+
+    handlers = cfg['handlers']
+
+    for handler in handlers.values():
+        filename = handler.get('filename', None)
+
+        if filename:
+            if not os.path.isabs(filename):
+                handler['filename'] = os.path.join(output_dir, filename)
+            log_files.append(handler['filename'])
+
+    return log_files
+
+
+def configure_logging(cfg_file: str = None,
+                      output_dir: str = None,
+                      console_log_level: str = None) -> list:
+    """Configure logging.
+
+    Parameters
+    ----------
+    cfg_file : str, optional
+        Logging config file. If `None`, defaults to `configure-logging.yml`
+    output_dir : str, optional
+        Output directory for the log files. If `None`, log only to the console.
+    console_log_level : str, optional
+        If `None`, use the default (INFO).
+
+    Returns
+    -------
+    log_files : list
+        Filenames that will be logged to.
+    """
+    if cfg_file is None:
+        cfg_loc = Path(esmvalcore.__file__ + "esmvalcore")
+        # TODO change to new location of config module in 2.3.0
```

Comment:
Still not too happy about this, but as long as you fix it when it breaks at the next release it should be fine :wink:

Reply:
promise :crossed_fingers: